Gems/Atom/RHI/WebGPU/Code/Source/RHI/RenderPipeline.cpp

```diff
@@ -0,0 +1,236 @@
+            wgpuDepthStencilState.stencilBack.passOp = ConvertStencilOp(stencilState.m_backFace.m_passOp);
+            wgpuDepthStencilState.stencilReadMask = stencilState.m_readMask;
+            wgpuDepthStencilState.stencilWriteMask = stencilState.m_writeMask;
+            wgpuDepthStencilState.depthBias = descriptor.m_renderStates.m_rasterState.m_depthBias;
+            wgpuDepthStencilState.depthBiasSlopeScale = descriptor.m_renderStates.m_rasterState.m_depthBiasSlopeScale;
+            wgpuDepthStencilState.depthBiasClamp = descriptor.m_renderStates.m_rasterState.m_depthBiasClamp;
+            m_wgpuRenderPipelineDescriptor.depthStencil = &wgpuDepthStencilState;
+        }
+        else
+        {
+            m_wgpuRenderPipelineDescriptor.depthStencil = nullptr;
+        }
+    }
+
+    void RenderPipeline::BuildMultisampleState(
+        const RHI::PipelineStateDescriptorForDraw& descriptor)
+    {
+        wgpu::MultisampleState& multisampleState = m_wgpuRenderPipelineDescriptor.multisample;
+        multisampleState.count = descriptor.m_renderStates.m_multisampleState.m_samples;
+        multisampleState.mask = 0xFFFFFFFF;
+        multisampleState.alphaToCoverageEnabled = descriptor.m_renderStates.m_blendState.m_alphaToCoverageEnable;
+    }
+
+    void RenderPipeline::BuildVertexState(const RHI::PipelineStateDescriptorForDraw& descriptor)
+    {
+        ShaderModule* module = BuildShaderModule(descriptor.m_vertexFunction.get());
+        if (module)
+        {
+            BuildConstants(
+                descriptor,
+                static_cast<const ShaderStageFunction*>(module->GetStageFunction())->GetSourceCode().data(),
+                m_vertexConstants);
+        }
+        wgpu::VertexState& vertexState = m_wgpuRenderPipelineDescriptor.vertex;
+        vertexState.module = module ? module->GetNativeShaderModule() : nullptr;
+        vertexState.entryPoint = module ? module->GetEntryFunctionName().c_str() : nullptr;
+        vertexState.constantCount = m_vertexConstants.size();
+        vertexState.constants = m_vertexConstants.empty() ? nullptr : m_vertexConstants.data();
+
+        const auto& buffers = descriptor.m_inputStreamLayout.GetStreamBuffers();
+        const auto& attributtes = descriptor.m_inputStreamLayout.GetStreamChannels();
+        vertexState.bufferCount = buffers.size();
+        vertexState.buffers = m_wgpuBuffers.data();
+        for (uint32_t i = 0; i < buffers.size(); ++i)
+        {
+            const auto& streamBuffer = buffers[i];
+            wgpu::VertexBufferLayout& bufferLayout = m_wgpuBuffers.emplace_back(wgpu::VertexBufferLayout{});
+            bufferLayout.arrayStride = streamBuffer.m_byteStride;
+            bufferLayout.stepMode = ConvertVertexStep(streamBuffer.m_stepFunction);
+            bufferLayout.attributes = m_wgpuAttributes.data() + m_wgpuAttributes.size();
+            for (uint32_t j = 0; j < attributtes.size(); ++j)
```

Comment:
maybe doing a find if would be easier to get the logic + would prevent iterating on all attributes and it would stop with first entry. It really depends if you need to have multpiple match or not